File: backend/app/api/v1/escrow.py
import uuid

from fastapi import APIRouter, Depends, Query

from app.api.deps import DbSession, requires
from app.models.escrow import EscrowState
from app.models.user import User
from app.rbac.permissions import Access, Module
from app.schemas.escrow import EscrowOut, EscrowPage
from app.services import escrow

router = APIRouter(prefix="/escrow", tags=["escrow"])

Manager = Depends(requires(Module.PAYMENT_ESCROW, Access.FULL))


@router.get("", response_model=EscrowPage)
async def list_escrows(
    session: DbSession,
    page: int = Query(1, ge=1),
    size: int = Query(25, ge=1, le=100),
    state: EscrowState | None = None,
    _: User = Manager,
) -> EscrowPage:
    items, total = await escrow.paginate(session, page, size, state)
    return EscrowPage(
        items=[EscrowOut.of(item) for item in items], total=total, page=page, size=size
    )


@router.get("/{escrow_id}", response_model=EscrowOut)
async def get_escrow(escrow_id: uuid.UUID, session: DbSession, _: User = Manager) -> EscrowOut:
    return EscrowOut.of(await escrow.get(session, escrow_id))


@router.post("/{escrow_id}/advance", response_model=EscrowOut)
async def advance_escrow(escrow_id: uuid.UUID, session: DbSession, _: User = Manager) -> EscrowOut:
    """Move the settlement one step forward. Reaching RELEASED pays the seller."""
    return EscrowOut.of(await escrow.advance(session, escrow_id))

File: frontend/lib/api/escrow.ts
import { apiClient } from "@/lib/api/client";
import type { Escrow, EscrowPage, ListEscrowsParams } from "@/types/escrow";

const BASE = "/api/v1/escrow";

export function listEscrows(accessToken: string, params: ListEscrowsParams = {}): Promise<EscrowPage> {
  const query = new URLSearchParams();
  if (params.page) query.set("page", String(params.page));
  if (params.size) query.set("size", String(params.size));
  if (params.state) query.set("state", params.state);

  const queryString = query.toString();
  return apiClient.get<EscrowPage>(`${BASE}${queryString ? `?${queryString}` : ""}`, { accessToken });
}

export function getEscrow(accessToken: string, escrowId: string): Promise<Escrow> {
  return apiClient.get<Escrow>(`${BASE}/${escrowId}`, { accessToken });
}

/** Moves the escrow one step forward. Reaching "released" pays the seller. */
export function advanceEscrow(accessToken: string, escrowId: string): Promise<Escrow> {
  return apiClient.post<Escrow>(`${BASE}/${escrowId}/advance`, undefined, { accessToken });
}